genai-perf/genai_perf/config/generate/search_parameters.py

```diff
@@ -0,0 +1,382 @@
+# Copyright 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from math import log2
+from typing import Any, Dict, List, Optional
+
+from genai_perf.config.generate.search_parameter import (
+    ParameterCategory,
+    ParameterList,
+    ParameterUsage,
+    SearchParameter,
+)
+from genai_perf.config.input.config_command import ConfigCommand, Range
+from genai_perf.exceptions import GenAIPerfException
+
+
+class SearchParameters:
+    """
+    Contains information about all configuration parameters the user wants to search
+    """
+
+    # These map to the run-config-search fields
+    # See github.com/triton-inference-server/model_analyzer/blob/main/docs/config.md
+    exponential_range_parameters = [
+        "model_batch_size",
+        "runtime_batch_size",
+        "concurrency",
+        "request_rate",
+    ]
+    linear_range_parameters = ["instance_count"]
+
+    model_parameters = [
+        "model_batch_size",
+        "instance_count",
+        "max_queue_delay",
+    ]
+    runtime_parameters = ["runtime_batch_size", "concurrency", "request_rate"]
+
+    def __init__(
+        self,
+        config: ConfigCommand = ConfigCommand(),
+        is_bls_model: bool = False,
+        is_ensemble_model: bool = False,
+        is_composing_model: bool = False,
+    ):
+        self._config = config.optimize
+
+        # TODO: OPTIMIZE
+        # self._supports_model_batch_size = model.supports_batching()
+        self._supports_model_batch_size = True
+
+        self._search_parameters: Dict[str, SearchParameter] = {}
+        self._is_ensemble_model = is_ensemble_model
+        self._is_bls_model = is_bls_model
+        self._is_composing_model = is_composing_model
+
+        self._populate_search_parameters()
+
+    ###########################################################################
+    # Accessor Methods
+    ###########################################################################
+    def get_parameter(self, name: str) -> Optional[SearchParameter]:
+        return self._search_parameters.get(name)
+
+    def get_type(self, name: str) -> ParameterUsage:
+        return self._search_parameters[name].usage
+
+    def get_category(self, name: str) -> ParameterCategory:
+        return self._search_parameters[name].category
+
+    def get_range(self, name: str) -> Range:
+        return Range(
+            min=self._search_parameters[name].min_range,  # type: ignore
```

Comment:
why do we need to ignore typing here?

Reply:
The underlying type is Optional[int], because the parameter could be a list instead of a range (in which case min/max_range would be None). But, if we are calling this method then you know that you are using a range and not a list. However, there's no way for me to communicate this to the static type checker...

Reply:
I feel like there should be a way, albeit probably very complex, to indicate this.
This is not a sticking point tho worth gating an approval.
Thanks for explaining it to me.